src/window.rs

```diff
@@ -0,0 +1,138 @@
+use core::marker::PhantomData;
+use core::f64;
+
+use {Sample, FloatSample};
+use signal;
+use signal::{Phase, ConstHz, Step};
+use conv::{FromSample, ToSample};
+use frame::Frame;
+
+// Using types instead of enum allows for static dispatch
+pub trait WindowType<S: Sample, T: Step> {
+    fn at_phase(phase: S) -> S;
+}
+
+pub struct HanningWindow;
+
+impl<S: Sample + ToSample<f64> + FromSample<f64>, T: Step> WindowType<S, T> for HanningWindow {
+    fn at_phase(phase: S) -> S {
+        let v = phase.to_sample::<f64>() * f64::consts::PI * 2.;
+        (0.5f64 * (1f64 - v.cos())).to_sample::<S>()
+    }
+}
+
+pub struct RectangleWindow;
+
+impl<S: Sample + FromSample<f64>, T: Step> WindowType<S, T> for RectangleWindow {
+    #[allow(unused)]
+    fn at_phase(phase: S) -> S {
+        (1.).to_sample::<S>()
+    }
+}
+
+pub struct Window<S: Sample, T: Step, F: Frame<Sample=S>, WT: WindowType<S, T>> {
+    pub phase: Phase<ConstHz>,
+    stype: PhantomData<S>,
+    ttype: PhantomData<T>,
+    ftype: PhantomData<F>,
+    wttype: PhantomData<WT>
+}
+
+impl<S: Sample, F: Frame<Sample=S>, WT: WindowType<S, ConstHz>> Window<S, ConstHz, F, WT> {
+    pub fn new(len: usize) -> Self {
+        let step = signal::rate(len as f64 - 1.).const_hz(1.);
+        Window {
+            phase: signal::phase(step),
+            stype: PhantomData,
+            ttype: PhantomData,
+            ftype: PhantomData,
+            wttype: PhantomData
+        }
+    }
+}
+
+impl<S: Sample + FromSample<f64>, F: Frame<Sample=S>, T: Step, WT: WindowType<S, T>> Iterator for Window<S, T, F, WT> {
+    type Item = F;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        let v = WT::at_phase((self.phase.next_phase()).to_sample::<S>());
+        Some(F::from_fn(|_| v))
+    }
+}
+
+pub struct WindowedFrame<'a, S: Sample + FromSample<f64>, T: Step, F: 'a + Frame<Sample=S>, WT: WindowType<S, T>> {
+    pub data: &'a [F],
+    pub window: Window<S, T, F, WT>,
+    idx: usize
+}
+
+impl<'a, S: Sample + FromSample<f64>, T: Step, F: 'a + Frame<Sample=S>, WT: WindowType<S, T>> WindowedFrame<'a, S, T, F, WT> {
+    pub fn new(data: &'a [F], window: Window<S, T, F, WT>) -> Self {
+        WindowedFrame {
+            data: data,
+            window: window,
+            idx: 0
+        }
+    }
+}
+
+impl<'a, S: Sample<Float=S> + FloatSample + FromSample<f64>, T: Step, F: 'a + Frame<Sample=S::Float>, WT: WindowType<S, T>> Iterator for WindowedFrame<'a, S, T, F, WT> {
```

Comment:
When I have more than one type with trait bounds, I normally find it much easier to distinguish the generics from their bounds if I move them to a where clause, I think most of these impl blocks would largely benefit from this, i.e.

```rust
impl<'a, S, T, F, WT> Iterator for WindowedFrame<'a, S, T, F, WT>
    where S: Sample<Float=S>,
          T: Step,
          F: 'a + Frame<Sample=S::Float>,
          WT: WindowType<S, T>,
```